### ISSUE TITLE
WO-024: trade matching

### ISSUE THREAD

**Issue by Xocas12** [Xocas12/gosplan-env#29]
> ## NOT ISSUABLE YET - THIS CARD IS A PLACEHOLDER AND ITS CONTENT IS PROVISIONAL
>
> **This work order is issued only AFTER the Phase-2 spec revision.** It is gated on gate **G2** and
> on the lead's spec revision that follows G2. Nothing here may be started before both have happened.
>
> **The behaviour of the P2 sketch is frozen at that revision, and not before.** PLAN §2.13 is
> labelled a *P2 sketch* and ends "Frozen at the P2 spec revision." The matching rule, the
> tie-breaking, the symmetrisation of the visibility mask and the exact surplus evaluation point are
> therefore undecided today; `gosplan/env/trade.py` holds signatures and constraints, not a rule.
>
> **Everything under "Implementation notes" below is PROVISIONAL** until the lead revises the spec.
> Do not read it as settled and do not present it to anyone else as settled.
>
> **An implementer handed this card in its present form stops and files an AMBIGUITY REPORT**
> (`workorders/AMBIGUITY_TEMPLATE.md`, CONTRACT rule 3).

> **WO-024** - Phase 2 - Owner: **LEAD writes the matching rule and the surplus definition in the spec revision; MID-strong implements** (PLAN §12.4, verbatim) - Difficulty **set by the LEAD at issue** - Gate **G3**
> Card: [`workorders/WO-024.md`](https://github.com/Xocas12/gosplan-env/blob/main/workorders/WO-024.md)

## The owner split, explicitly

PLAN §12.4 splits this card: **the LEAD writes the matching rule and the surplus definition into the
spec revision; a MID-strong session implements what the lead wrote.** These are two different pieces
of work and the second cannot begin before the first is on disk. An implementer session that finds
the matching rule still undecided has been handed the card too early: it files an ambiguity report
and stops. It does not design the matcher.

## What this is

Lets enterprises trade inputs with each other directly, outside the plan. Each enterprise emits a
per-good offer in `[−1, 1]`: positive offers a fraction of its own input stock, negative asks for a
fraction of what it needs. A visibility mask decides who can see whom, a greedy matcher pairs
complementary offers at plan-price parity, and a per-unit transaction cost destroys some of what
moves. This is the mechanism behind *blat* - the informal horizontal exchange that kept Soviet
enterprises supplied when the plan under-delivered. The mechanism is here; the blat measurement
(PLAN §4.1 row 6) is held out and computed for the first time in WO-030.

## Depends on

- gate G2 - the Phase-1 gate (PLAN §4.5 criteria 1-4, signed off Human + LEAD). This card cannot be
  started before G2 has passed.
- the **P2 spec revision** (LEAD), which for this card means specifically the matching rule and the
  surplus definition.
- WO-013
- WO-004
- WO-005
- WO-007
- WO-008
- WO-009

Trade sits after DELIVER in the period schedule (PLAN §2.5). Consumed by WO-030 (phenomenon row 6,
blat) and WO-031 (the P2 acceptance run).

**Blocked by:** #18 (WO-013), #9 (WO-004), #10 (WO-005), #12 (WO-007), #13 (WO-008), #14 (WO-009), #45 (gate G2)
This card may not be started until every issue above is closed (CONTRACT rule 12 / the loop in CONTRIBUTING.md).

## Read first (exhaustive whitelist)

- `PLAN.md` §§ 2.3, 2.4, 2.5, 2.6, 2.9.1, 2.13, 2.15, 3, 4.2, 4.4 - PLAN.md is git-ignored and
  supplied out of band; it is not in the repository and has no blob link.
- [`CONTRACT.md`](https://github.com/Xocas12/gosplan-env/blob/main/CONTRACT.md) - rules 4, 6, 7, 9
- [`spec/spec.py`](https://github.com/Xocas12/gosplan-env/blob/main/spec/spec.py) -
  `EnterpriseAction.trade_offer`, `match_trades`, `enterprise_reward`, `Purpose`
- [`spec/CHANGELOG.md`](https://github.com/Xocas12/gosplan-env/blob/main/spec/CHANGELOG.md)
- [`gosplan/config.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/config.py)
- [`gosplan/params.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/params.py)
- [`gosplan/rng.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/rng.py)
- [`gosplan/env/trade.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/trade.py)
- [`gosplan/env/production.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/production.py)
- [`gosplan/env/obs.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/obs.py)
- [`gosplan/env/reward.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/reward.py)
- [`gosplan/env/state.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/state.py)
- [`gosplan/env/step.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/step.py)
- [`docs/ref_worked_example.md`](https://github.com/Xocas12/gosplan-env/blob/main/docs/ref_worked_example.md)
- **the revised spec text, which does not exist yet - that is what gates this card**

Read nothing else. CONTRACT rule 12.

## Write only

Provisional; the revision fixes the final list.

- `gosplan/env/trade.py` - `match_trades`, `visible_counterparties`, `trade_surplus`
- `gosplan/env/obs.py` - the horizontal observation block, gated by `horizontal_visibility`
- `gosplan/env/step.py` - the trade stage hook, if the revision moves it

Not this card: `ref/`, `tests/**`, `spec/spec.py`. Only `match_trades` is signature-frozen; the other
two functions in `gosplan/env/trade.py` are internal and may be restructured by the revision (PLAN
§0, finding F14).

Creating, editing or deleting any other file is a CONTRACT rule 12 violation.

## Must pass

**No new test ids are frozen yet** - the revision writes them and names their ids in this section
(CONTRACT rule 12).

Already binding and already determinate, green after this change:

- [ ] **T-U1** conservation to 1e-9, with the traded quantities and the `τ` loss as explicit terms of
      the per-period per-good identity - the goods `τ` destroys must not simply vanish -
      `tests/unit/test_conservation.py`
- [ ] **T-B1** no hard-coded pathology: `TruthfulMyopic` executes **no trade** -
      `tests/behavioural/test_no_hardcoded_pathology.py`
- [ ] **T-B5** welfare blindness: the horizontal block may carry only other enterprises' last report
      ratios in the own sector (PLAN §2.4), never their `y`, `S` or `X` -
      `tests/behavioural/test_welfare_blindness.py`
- [ ] **T-B7** golden parity with `ref/` - `tests/golden/test_golden_parity.py`

Acceptance criterion PLAN already fixes: **at `horizontal_visibility = 0.0` (the Phase-1 default) the
visibility mask is empty for every enterprise, `match_trades` returns immediately, surplus is
identically zero and the observation gains no field**, so every golden file passes unchanged.

Completion command, copy-paste:

```bash
uv run ruff check . && uv run ruff format --check . && uv run pytest -q
```

The revision appends the specific `pytest tests/unit/<new test file> -q` invocation. `pytest -q`
never collects `tests/acceptance/` (CONTRACT rule 13).

## Implementation notes

**Provisional until the lead revises the spec.**

### Locked Phase-2 values (PLAN §4.2, row "6 Blat")

Locked **now**:

```
horizontal_visibility = 1.0
trade_tau        τ    = 0.05
```

### Offer semantics (PLAN §2.13)

`trade_offer` is a per-good action component in `[−1, 1]` (PLAN §2.3):

```
trade_offer_ij > 0   offers up to that fraction of X_ij      (own input stock)
trade_offer_ij < 0   wants  up to that fraction of need_ij   (this period's requirement)
```

The two sides are scaled by **different quantities on purpose** - stock for offers, need for wants -
so an offer is bounded by what you physically have and a want is bounded by what you physically
lack. Do not "simplify" them onto a common scale.

### Visibility

A random subset of the other `N − 1` enterprises, of size `horizontal_visibility · (N − 1)`, drawn
with purpose `trade_visibility` through `gosplan.rng.draw` (CONTRACT rule 9, PLAN §2.15), **redrawn
per period**. Visibility is an information parameter, not a fixed network: a fixed counterparty
graph would turn an information treatment into a topology treatment and is explicitly forbidden.

At `horizontal_visibility = 0.0` the mask is empty for every enterprise; at `1.0` (the locked
Phase-2 value) everyone sees everyone.

### Matching and price

Greedy by largest complementary pairs, executed at **plan-price parity**:

```
p_s · Q_s = p_b · Q_b        with p from state.plan_prices
```

with a per-unit transaction cost `τ` on the executed quantity, which **physically disappears** -
i.e. it is destroyed, not transferred, and it must appear as an explicit loss term of the T-U1
identity (tolerance 1e-9 per period per good).

### Surplus, verbatim (PLAN §2.13)

```
Δh_i = p_{s(i)} · [ Ŷ_i(X_after) − Ŷ_i(X_before) ]
```

where `Ŷ_i` is next step's **intended output at the agent's last effort**, and it is **computed by
the environment, never self-reported**. Two properties this construction buys, both of which are
requirements:

- wash trades are unprofitable, because `τ > 0` and `Δh` is the same function of the same true state
  for both sides. A surplus definition under which a round trip pays is a defect;
- `trade_surplus` is the single admitted trade term in the reward (CONTRACT rule 4:
  `scale · (B(ρ) − 1[audited] · Pen + trade_surplus)` at the report step). There is no volume bonus,
  no counterparty-count bonus, no reciprocity bonus, and no self-reported gain.

### What `match_trades` may not touch

Trade **reallocates inputs**. It does not manufacture output, forgive a target or settle a claim, so
`match_trades` may not write `inv_output`, `target`, any `last_*` field, `cum_output` or `cum_cost`.

### What PLAN does NOT decide - the revision must state each before this card is issued

- **the tie-breaking rule of the greedy matcher**, and the iteration order that makes it
  deterministic and reproducible in `ref/ref_step.py`. This is load-bearing twice over: `ref/` is the
  golden-file source (T-B7), and WO-029's JAX port must reproduce the same tie-break for its 1e-5
  parity criterion to mean anything.
- **whether the visibility mask is symmetrised, and how.** PLAN says a trade needs both sides to see
  each other; the rule that turns two independent draws into a symmetric relation is not written.
- **whether `τ` is charged in units of the good moved, on both legs or one, and to whom.**
- **the exact step at which `Ŷ` is evaluated.** "Next step's intended output at the agent's last
  effort" fixes the effort but not the coverage state used for `need`.
- **what happens at the last PRODUCE step of a period**, where "next step" is the REPORT step.

Do not pick any of these. CONTRACT rule 3: choosing "the reasonable default" is a violation.

## Forbidden

- Computing, plotting or testing PLAN §4.1 rows **2, 5, 6, 7** - held out. Row 6 (blat) is this
  mechanism's own phenomenon and is computed for the first time in WO-030, inside the P2 acceptance
  run of WO-031 only.
- Any reward term for trade other than the single `trade_surplus` of CONTRACT rule 4: no bonus for
  volume, for counterparty count, for reciprocity, and no self-reported gain.
- Letting `match_trades` write `inv_output`, `target`, any `last_*` field, `cum_output` or
  `cum_cost`.
- Any rule that makes enterprises trade, or that prices a trade so that trading is mechanically
  profitable. Blat must be a consequence of binding coverage plus visibility (CONTRACT rule 7).
- Direct `numpy.random` / `jax.random` under `gosplan/env/` (CONTRACT rule 9); a fixed counterparty
  network in place of the per-period keyed draw.
- Changing the PLAN §4.2 locked values (`horizontal_visibility = 1.0`, `τ = 0.05`).
- Designing the matching rule or the surplus definition from an implementer session - that is the
  LEAD's half of this card.

## Contract rules that bite here

- **Rule 3 - STOP AND REPORT.** Five questions above are undecided, and the matching rule itself is
  LEAD work. Choosing "the reasonable default" is a violation; file an ambiguity report and stop.
- **Rule 12 - work orders.** Read only the whitelist, write only the listed files, run the
  completion command verbatim, report in the four-part format.
- **Rule 4 - reward terms.** `trade_surplus` is the only admitted trade term, and it is the
  environment's number, not the agent's.
- **Rule 6 - welfare blindness.** The horizontal observation block carries other enterprises' last
  report ratios in the own sector and nothing else - no `y`, no `S`, no `X`, no welfare.
- **Rule 7 - no hard-coded pathology.** Nothing may make agents trade or make trading mechanically
  profitable.
- **Rule 9 - RNG.** The visibility mask is a per-period keyed draw with purpose `trade_visibility`,
  never a fixed network and never a global generator.
- **Rule 2 - frozen tests.** T-B1's "TruthfulMyopic executes no trade" clause is frozen; if it looks
  wrong, file an ambiguity report.

## Definition of done

- [ ] Every must-pass test green, including the frozen Phase-1 suite unchanged
- [ ] At `horizontal_visibility = 0.0` `match_trades` returns immediately, surplus is identically
      zero, the observation gains no field, and every golden file passes unchanged
- [ ] Traded quantities and the `τ` loss appear as explicit terms of the T-U1 identity
- [ ] `uv run ruff check . && uv run ruff format --check .` clean
- [ ] `uv run python scripts/contract_guard.py --all` clean
- [ ] No file outside "Write only" created, edited or deleted
- [ ] CI `ci-ok` green on the PR
- [ ] The four-part report posted as the PR body: 1) tests passed/failed with output, 2) ambiguity
      reports filed, 3) files changed, 4) anything you were tempted to invent and did not
